File: NFMWorld.Library/Gamemodes/Lua/LuaStage.cs
using nfm_world_library.Lua;
using NFMWorldLibrary.Backend;

namespace NFMWorldLibrary.Gamemodes.Lua;

/// <summary>
/// Lua-facing view over the stage being raced. Exposes the small surface
/// gamemode scripts need without leaking the full backend stage graph.
/// </summary>
[LuaVisible]
public sealed partial class LuaStage
{
    private readonly BackendStage _backend;

    [LuaHidden]
    public LuaStage(BackendStage backend) => _backend = backend;

    [LuaName("name")]
    public string Name => _backend.Path;

    [LuaName("nlaps")]
    public int Nlaps => _backend.nlaps;

    [LuaName("checkpointCount")]
    public int CheckpointCount => _backend.checkpoints.Count;

    [LuaName("checkpointPosition")]
    public f64Vector3 CheckpointPosition(int index) => _backend.checkpoints[index].Position;
}

File: NFMWorld.Library/data/lua/library/LuaStage.lua
---@class LuaStage
---@field name string
---@field nlaps integer
---@field checkpointCount integer
---@field checkpointPosition fun(self: LuaStage, index: integer): fixed64vector3

LuaStage = {}


---Creates a new LuaStage
---@return LuaStage
function LuaStage.new() end
